File: generators/zml/primitives/lm_head.zig
//! lm_head@1.0.0 — one logit per vocabulary entry.

const p = @import("../primitive.zig");

pub const primitive: p.Primitive = .{
    .name = "lm_head",
    .version = "1.0.0",
    .run = run,
    .capabilities =
    \\{"arguments": {"width": "any", "vocabulary": "any"}, "states": []}
    ,
};

fn run(ctx: *p.Ctx, call: p.Call) ![]const p.Binding {
    return p.one(ctx, "logits", p.linear(call.inputs.must("input"), call.params.must("weight")));
}

File: generators/onnx/primitives/lm_head.py
"""lm_head@1.0.0 — one logit per vocabulary entry; `softcap` refused."""
from primitives._common import linear, supports_from

PRIMITIVE = ("lm_head", "1.0.0")
CAPABILITIES = {"arguments": {"width": "any", "vocabulary": "any", "softcap": "absent"}, "states": []}


def supports(arguments):
    return supports_from(CAPABILITIES, arguments)


def emit(ctx, arguments, inputs, params, states):
    return {'logits': linear(ctx, inputs['input'], ctx.param(params['weight']))}
